src/utils/uploader.js

```diff
@@ -26,64 +25,94 @@ class Uploader {
-    }
-
-    async #initType(rpc, contractAddress, uploadType) {
-        const provider = new ethers.JsonRpcProvider(rpc);
-        const fileContract = new ethers.Contract(contractAddress, FlatDirectoryAbi, provider);
-        let isSupportBlob;
         try {
-            isSupportBlob = await fileContract.isSupportBlob();
+            this.#flatDirectory = await FlatDirectory.create({
+                rpc: rpc, privateKey: pk, address: contractAddress
+            });
         } catch (e) {
-            Logger.error(`Failed to initialize upload type for contract ${contractAddress}. ${e.message}`);
+            if (e.message.includes('The current SDK does not support this contract')) {
+                Logger.error("Failed to query contract. The contract was created with ethfs-cli 1.x and is not compatible with version 2.0. Please switch back to ethfs-cli 1.x to access this contract.");
+            } else {
+                Logger.error(`SDK initialization failed, Please check your parameters and network connection, and try again.  info=${e.message}`);
+            }
             return false;
         }
 
         if (uploadType) {
             // check upload type
-            if (!isSupportBlob && Number(uploadType) === UPLOAD_TYPE_BLOB) {
+            if (!this.#flatDirectory.isSupportBlob && uploadType === UploadType.Blob) {
                 Logger.error(`Network does not support this upload type. Please switch to another type. Type=${uploadType}`);
                 return false;
             }
-            this.#uploadType = Number(uploadType);
+            this.#uploadType = uploadType;
         } else {
-            this.#uploadType = isSupportBlob ? UPLOAD_TYPE_BLOB : UPLOAD_TYPE_CALLDATA;
+            this.#uploadType = this.#flatDirectory.isSupportBlob ? UploadType.Blob : UploadType.Calldata;
         }
         return true;
     }
 
+    async #fetchFileDataBatch(fileBatch) {
+        const keys = fileBatch.map(file => file.name);
+        const fileHashArr = await this.#flatDirectory.fetchHashes(keys);
+        return fileBatch.map(file => {
+            file.chunkHashes = fileHashArr[file.name];
+            return file;
+        })
+    }
+
+    #getFileChunkCount(fileSize) {
+        if (this.#uploadType === UploadType.Blob) {
+            return Math.ceil(fileSize / OP_BLOB_DATA_SIZE);
+        } else {
+            if (fileSize > 24 * 1024 - 326) {
+                // Data need to be sliced if file > 24K
+                return Math.ceil(fileSize / (24 * 1024 - 326));
+            }
+            return 1;
+        }
+    }
+
+    #groupFiles(files) {
+        return from(files).pipe(
+            scan((acc, file) => {
+                const newFiles = [...acc.files, file];
+                const totalChunks = acc.totalChunks + this.#getFileChunkCount(file.size);
+                const totalFileCount = acc.totalFileCount + 1;
+
+                if (totalChunks >= MAX_CHUNKS || totalFileCount === files.length) {
```

Comment:
why return empty result when `totalFileCount === files.length`?

Reply:
Ensure that the last part of files that do not meet the number of chunks also become a group.